GH-46128:[C++] Add CompactArray method to BinaryViewArray types

andishgar opened the pull request:
### Rationale for this change
As we discussed [here](https://github.com/apache/arrow/issues/46128#issuecomment-2820665771), the following method copies only data that is used in data buffers to a new buffer
### What changes are included in this PR?
Add a new method the name of which is `CompactArray`
### Are these changes tested?
I run the relevant unit tests
### Are there any user-facing changes?
Yes, I add `ComapctArray` method to `BinaryViewArray` class
* GitHub Issue: #46128

andishgar commented:
@pitrou When I suggested adding these methods, I wasn't aware that`arrow::BinaryViewBuilder::AppendArraySlice `already serves the same purpose. What do you recommend I do? Should I drop this PR?

pitrou commented:
> When I suggested adding these methods, I wasn't aware that`arrow::BinaryViewBuilder::AppendArraySlice `already serves the same purpose. What do you recommend I do?

I still think it's a useful utility method to have. cc @bkietz for additional opinions.

> I found that the following code results in an error when using arrow::BinaryViewBuilder::AppendArraySlice. Is there any priority or plan to address this?

I don't think this bug is already reported, so there is no "priority or plan". Can you open a separate issue for it, and perhaps submit a PR if you're motivated?

andishgar commented:
> I don't think this bug is already reported, so there is no "priority or plan". Can you open a separate issue for it, and perhaps submit a PR if you're motivated?

Thank you for your response. Yes, of course

pitrou commented on cpp/src/arrow/array/array_binary_test.cc:
```diff
@@ -1007,11 +1013,97 @@ class TestBaseBinaryDataVisitor : public ::testing::Test {
+    ASSERT_EQ(compacted_array->data()->buffers.size(), 2);
+    AssertArraysEqual(*array, *compacted_array);
+
+    auto buffer_a = Buffer::FromString(std::string(1 << 5, 'a'));
```
Can this be less cryptic? `1 << 5` can just be written `32`.

pitrou commented on cpp/src/arrow/array/array_binary_test.cc:
```diff
@@ -1007,11 +1013,97 @@ class TestBaseBinaryDataVisitor : public ::testing::Test {
+                   {buffer_a, buffer_b},
+                   {
+                       util::ToBinaryView(
+                           "bbbb", static_cast<int32_t>(buffer_b->size() >> 1), 1, 16),
```
`static_cast<int32_t>(buffer_b->size() >> 1)` is pointlessly cryptic, how about 16 or something?

pitrou commented on cpp/src/arrow/array/array_binary.cc:
```diff
@@ -104,7 +105,14 @@ BinaryViewArray::BinaryViewArray(std::shared_ptr<DataType> type, int64_t length,
+  ARROW_RETURN_NOT_OK(MakeBuilder(pool, type(), &builder));
+  ArraySpan span(*data());
+  // ArraySpan handles offset
+  ARROW_RETURN_NOT_OK(builder->AppendArraySlice(span, 0, span.length));
```
Hmm, ok, but this will always copy the data even if it is already compact? Is there a way to avoid that by looking at the buffers directly?

pitrou replied:
> I'm currently reviewing the implementation, and I think we should consider making some changes:
> 
>     1. **Bitmap Buffer**:
>        It's possible to avoid copying the bitmap buffer entirely.

Yes, certainly, since the array contains the same data logically (beware of the array `offset`, though).

>     2. **View Buffer**:
>        For the view buffer, we likely need to make a copy, since it may already be shared due to operations like `cast`, `arrow::Array::View`, or `arrow::ViewOrCopyTo`.
>        However, I wonder if we could use `std::shared_ptr::use_count()` to detect whether the array is the sole owner of the view buffer and skip the copy in that case.

We certainly should not do that. Array data is immutable in Arrow. Even if the `use_count` is 1, the data may be in use in another thread, or a raw pointer may have been obtained somewhere.

>     3. **Data Buffer**:
>        Three scenarios come to mind:
>        3.1. The buffer is entirely empty — in such cases, we can safely discard it. This might occur after calling a compute function like `extract_regex` (note: this is a hypothetical example, as compute functions don’t support `StringView` yet).
>        3.2. Only a small portion of the data buffer is used — for instance, if only one-tenth of the buffer is referenced by the view. In this case, copying the used portion to a new buffer helps prevent memory bloat.
>        3.3. A significant portion of the buffer (e.g., 50% or more) is referenced — in such a case, it may be better to preserve the original data buffer, unless we're specifically trying to reduce memory usage.

We could certainly expose a parameter to choose the ratio of unused buffer space above which to compact a buffer.

> Additionally, there's another concern: it might be useful to expose a public API method to indicate whether calling `CompactArray` would be beneficial.

Well, the decision is buffer-grained, not array-grained: some buffers might be worth compacting and some others not.

> What do you think? Should I integrate this heuristic into `CompactArray`, or would it be better to split it into a separate method or even into a different pull request?

As said above, I think it would make sense to integrate the heuristic and apply it at the buffer level.

pitrou replied:
> Pardon me, just to be sure — does your overall suggestion correspond to pseudocode like this?

Yes, indeed.

pitrou commented on cpp/src/arrow/array/array_binary_test.cc:
```diff
@@ -1007,11 +1013,97 @@ class TestBaseBinaryDataVisitor : public ::testing::Test {
+
+ public:
+  void SetUp() override { type_ = TypeTraits<Type>::type_singleton(); }
+  void TestType() {
```
Nit: can you leave empty lines between methods?
```suggestion

  void TestType() {
```

pitrou commented on cpp/src/arrow/array/array_binary_test.cc:
```diff
@@ -1007,11 +1013,97 @@ class TestBaseBinaryDataVisitor : public ::testing::Test {
+  std::shared_ptr<DataType> type_;
+};
+
+// Should We move this to gtest_util.h
```
Please do!

pitrou reviewed (changes requested):
Thanks for posting this PR and sorry for the delay @andishgar . I added some comments below.

andishgar commented:
@pitrou I’ve implemented an initial prototype, and I believe a few points are worth mentioning.

First, here is [my initial solution](https://github.com/user-attachments/files/21074918/1493a4.zip).
Please note that the code is currently quite messy ("spaghetti code") and lacks any testing, so it likely contains logical errors. My goal with this prototype was not correctness but rather to explore what might be required to implement this feature.

1. There is a function [here](https://github.com/apache/arrow/blob/66ee781dc2c7964950c7066d139fd592ac5700e5/cpp/src/arrow/compute/kernels/util_internal.h#L158-L175) for retrieving a bitmap buffer without copying, when possible. I suggest we open a separate pull request to move this function to [arrow/util/bitmap\_ops.h](https://github.com/apache/arrow/blob/66ee781dc2c7964950c7066d139fd592ac5700e5/cpp/src/arrow/util/bitmap_ops.h#L61), and then continue work on this pull request. What is your opinion?
2. To track which parts of a data buffer are used, I plan to use [boost::icl::interval_set](https://www.boost.org/doc/libs/1_88_0/libs/icl/doc/html/index.html#boost_icl.introduction.interval_combining_styles). While I’ve written an initial version (corresponding to the `Add` function in the uploaded code), I’m not entirely confident in its correctness.  I want to first open an issue and a pull request to port the `boost::icl::interval_set` data structure to arrow/util, along with its implementation and tests.(Note that this is my first time using Boost, so it may take some time.) I think this data structure might be useful in other parts of the codebase as well—possibly in compute
What do you think? Do you suggest a better algorithm or data structure for this use case?

By the way, I’m aiming to use the boost::icl::interval_set data structure in a way that enables an API like this:
```c++
#include <iostream>
#include <boost/icl/interval_set.hpp>

void Run() {
    boost::icl::interval_set<int>::type set;

    boost::icl::interval<int>::type interval_1(10, 15);
    boost::icl::interval<int>::type interval_2(2, 3);
    boost::icl::interval<int>::type interval_3(1, 5);
    set.add(interval_2);
    set.add(interval_1);
    set.add(interval_3);
    set.add(interval_1);

    auto it =set.find(3);
    std::cout << it->lower() << std::endl;

    for (auto it : set) {
        std::cout << it << std::endl;
    }

}

int main() {
    Run();
    return 0;
}

```

andishgar committed:
implement occupancy threshold

andishgar committed:
fix warnings

andishgar commented:
@pitrou I’ve applied your suggestion. The changes are ready for review.

pitrou commented on cpp/src/arrow/array/array_binary.cc:
```diff
@@ -105,6 +111,392 @@ BinaryViewArray::BinaryViewArray(std::shared_ptr<DataType> type, int64_t length,
+  }
+}
+
+struct Interval {
```
The interval code could probably be moved to a new header `arrow/util/interval_internal.h`.

pitrou replied:
And it will help writing unit tests for it.

andishgar replied:
Regarding writing a solution for this, I came up with something a bit unusual.

Every idea that came to mind, I tested using a kind of "glass-box" approach, writing tests with 100% path coverage — and they all passed. But eventually, I discovered a single interval that broke everything.

So in the end, my solution was to first generate two CSV files: one containing the original intervals and the other containing the merged intervals, calculated using a reliable library. Then, I added these intervals to my code and compared my results with the expected merged intervals. (I tested this on 100 million intervals, with a total size of about 1.6 GB.)

Would it make sense to import smaller CSV files and write tests based on those? Or should I just stick to writing unit tests to cover different paths?

pitrou commented on cpp/src/arrow/array/array_binary.cc:
```diff
@@ -105,6 +111,392 @@ BinaryViewArray::BinaryViewArray(std::shared_ptr<DataType> type, int64_t length,
+      JoinRight(it);
+    } else {
+      if (it->end < interval.end) {
+        const_cast<int64_t&>(it->end) = interval.end;
```
Why do you need a `const_cast`? The `insert` call returns a non-const iterator.

pitrou replied:
Ah, I see that iterators on `std::set` are always const. Then I would suggest making the `end` and `offset` members mutable.

pitrou commented on cpp/src/arrow/array/array_binary.cc:
```diff
@@ -105,6 +111,392 @@ BinaryViewArray::BinaryViewArray(std::shared_ptr<DataType> type, int64_t length,
+  };
+
+  bool ValidateOccupancyThreshold(double occupancy_threshold) {
+    return std::signbit(occupancy_threshold) || std::isnan(occupancy_threshold) ||
+           occupancy_threshold > 1;
```
I'm curious, how did you come to such a convoluted formula?

andishgar replied:
I think you mean I should document the behavior of the code that returns true when the value is negative, NaN, or greater than one. Am I right?
(My inspiration for this code comes from several pull requests in Arrow that I was involved in, related to floating-point comparisons.)

pitrou replied:
No, it just seems that this code is very convoluted for what it's trying to do.

pitrou commented on cpp/src/arrow/array/array_binary.cc:
```diff
@@ -105,6 +111,392 @@ BinaryViewArray::BinaryViewArray(std::shared_ptr<DataType> type, int64_t length,
+  IntervalSet interval_set;
+};
+
+class CompactArrayImpl {
```
I will defer on reviewing the rest of the implementation until the comments on the interval code are addressed.

andishgar commented on cpp/src/arrow/array/array_binary.cc:
```diff
@@ -105,6 +111,392 @@ BinaryViewArray::BinaryViewArray(std::shared_ptr<DataType> type, int64_t length,
+struct Interval {
+  int64_t start;
+  int64_t end;
+  int32_t offset = -1;
```
aybe I didn’t choose the best name—let me explain what "offset" means here.
Imagine the following intervals:
```css
[0,15] [10,30] [80,100] [90 120] [1000 100] [1090,200]
``` 
After adding and merging them, they become:
```css
[0,30] [80,120] [1000 1200]
``` 
The offsets for these intervals would then be:
```css
0,30,70
``` 
In other words, the offset represents the position of each merged interval as if all intervals were compacted together into a new buffer.

Do you think I should rename "offset" to something clearer, or is it understandable as is?

pitrou replied:
You could probably rename it to something clearer.
Moreover, I'm not sure it needs to live in the `Interval` structure, either. You only compute the offsets at the end, right?

andishgar commented on cpp/src/arrow/array/array_binary.cc:
```diff
@@ -105,6 +111,392 @@ BinaryViewArray::BinaryViewArray(std::shared_ptr<DataType> type, int64_t length,
+  int64_t CalculateOffsetAndTotalSize() {
+    int64_t total_size = 0;
+    for (auto& it : interval_set) {
+      const_cast<int32_t&>(it.offset) = static_cast<int32_t>(total_size);
```
Regarding overflow handling, there is no issue—I’ve written a test for the overflow case, and it passes.

I believe it's possible to reason why `static_cast<int32_t>` is safe in this context.

The intervals in view elements are determined by their `size` and `offset`, and neither of these values can exceed `INT32_MAX`. Therefore, the maximum range of data that can be referenced from view elements into a buffer is approximately 4 GB (minus 2 bytes).

Since the size attribute is limited to `INT32_MAX,` it is not possible to have an `occupancy` greater than `INT32_MAX`, unless there is an interval that includes or starts at the position `INT32_MAX.`


As I mentioned above,`Interval.offset `refers to the position of the interval in an imaginary compacted buffer. Since intervals must either begin at or include the position `INT32_MAX `in such a situation, the value never exceeds` INT32_MAX`.

What’s your opinion? Is my observation correct, or is there some exception?

andishgar commented on cpp/src/arrow/array/array_binary.cc:
```diff
@@ -105,6 +111,392 @@ BinaryViewArray::BinaryViewArray(std::shared_ptr<DataType> type, int64_t length,
 
+namespace {
+
+// TODO Should We move this to bitmap_ops.h and Remove from compute/kernels/util.s
+Result<std::shared_ptr<Buffer>> GetOrCopyNullBitmapBuffer(const ArrayData& in_array,
+                                                          MemoryPool* pool) {
```
Could I do this in a separate pull request? It seems that at least two compute kernels rely on this function ([cpp/src/arrow/compute/kernels/util_internal.h](https://github.com/apache/arrow/blob/b062998fdae7877686cd25b9f4d56db6dbff65d5/cpp/src/arrow/compute/kernels/util_internal.h#L158)). From my search on GitHub, I found that at least two compute kernels depend on it.

andishgar committed:
Implement fast Merge and other things

relocate GetOrCopyNullBitmapBuffer
reclocate IntervalMerger to util/interval.h
Write tests to cover all branches and lines of  Interval Merger

andishgar committed:
Fix typo
